pandas/tests/test_timedelta64_overflow.py

```diff
@@ -0,0 +1,322 @@
+    """
+    if issubclass(type_, (Timedelta, Timestamp)):
+        return type_(value)
+
+    if issubclass(type_, ExtensionArray):
+        box_cls = array
+    elif issubclass(type_, Index):
+        box_cls = Index
+    else:
+        box_cls = type_
+
+    return type_(tm.box_expected([value], box_cls))
+
+
+@given(positive_td=positive_tds)
+def test_add_raises_expected_error_if_result_would_overflow(
+    add_sub_types: BinaryOpTypes,
+    positive_td: Timedelta,
+):
+    left = wrap_value(Timedelta.max, add_sub_types.left)
+    right = wrap_value(positive_td, add_sub_types.right)
+
+    if add_sub_types.result is Timedelta:
+        msg = "|".join(
+            [
+                "int too big to convert",
+                "Python int too large to convert to C long",
+            ]
+        )
+    else:
+        msg = "Overflow in int64 addition"
+
+    with pytest.raises(OverflowError, match=msg):
+        left + right
+
+    with pytest.raises(OverflowError, match=msg):
+        right + left
+
+
+@xfail_no_overflow_check
+@given(positive_td=positive_tds)
+def test_sub_raises_expected_error_if_result_would_overflow(
+    add_sub_types: BinaryOpTypes,
+    positive_td: Timedelta,
+):
+    left = wrap_value(Timedelta.min, add_sub_types.left)
+    right = wrap_value(positive_td, add_sub_types.right)
+
+    msg = "Overflow in int64 addition"
+    with pytest.raises(OverflowError, match=msg):
+        left - right
+
+    with pytest.raises(OverflowError, match=msg):
+        (-1 * right) - abs(left)
+
+
+@given(td_value=positive_tds)
+def test_add_timestamp_raises_expected_error_if_result_would_overflow(
+    ts_add_sub_types: BinaryOpTypes,
+    td_value: Timedelta,
+):
+    left = wrap_value(td_value, ts_add_sub_types.left)
+    right = wrap_value(Timestamp.max, ts_add_sub_types.right)
+
+    ex = (OutOfBoundsDatetime, OverflowError)
+    msg = "|".join(["Out of bounds nanosecond timestamp", "Overflow in int64 addition"])
+
+    with pytest.raises(ex, match=msg):
+        left + right
+
+    with pytest.raises(ex, match=msg):
+        right + left
+
+
+@xfail_no_overflow_check
+@given(td_value=positive_tds)
+def test_sub_timestamp_raises_expected_error_if_result_would_overflow(
+    ts_add_sub_types: BinaryOpTypes,
+    td_value: Timedelta,
+):
+    right = wrap_value(td_value, ts_add_sub_types[0])
+    left = wrap_value(Timestamp.min, ts_add_sub_types[1])
+
+    ex = (OutOfBoundsDatetime, OverflowError)
+    msg = "|".join(["Out of bounds nanosecond timestamp", "Overflow in int64 addition"])
+
+    with pytest.raises(ex, match=msg):
+        left - right
+
+
+@given(value=st.floats().filter(lambda f: abs(f) > 1))
+def test_scalar_multiplication_raises_expected_error_if_result_would_overflow(
+    value: float,
+):
+    td = Timedelta.max
+
+    msg = "|".join(
+        [
+            "cannot convert float infinity to integer",
+            "Python int too large to convert to C long",
+            "int too big to convert",
```

Comment:
in most of these cases we probably want these to raise OutOfBoundsTImedelta with a timedelta-specific exception message

Reply:
That seems reasonable. AFAICT, the key to getting there is to catch `OverflowError`s here: https://github.com/pandas-dev/pandas/blob/9f2491875945d4b515f66491dbe507632060d6bf/pandas/_libs/tslibs/timedeltas.pyx#L264

Once that's one, I *think* we'd remove this check, too: https://github.com/pandas-dev/pandas/blob/9f2491875945d4b515f66491dbe507632060d6bf/pandas/core/nanops.py#L378-L380

Does that seem right, or am I missing something? I can open a follow-up PR for that work.